File: Data Structures & Algorithms/permutations/submission-5.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        res = [] 

        def dfs(part, track): 
            if len(part) == len(nums): 
                res.append(part.copy()) 
                return 

            for key in track: 
                if track[key]: 
                    continue 

                part.append(key) 
                track[key] = True 
                dfs(part, track) 
                part.pop() 
                track[key] = False  

        

        dfs([], {i:False for i in nums}) 
        return res

File: Data Structures & Algorithms/palindrome-partitioning/submission-5.js
class Solution {
    /**
     * @param {string} s
     * @return {string[][]}
     */
    partition(s) { 
        const res = [] 
        const part = [] 
        this.dfs(0, s, part, res) 
        return res
    } 

    dfs(start, s, part, res) {
        if (start === s.length) {
            res.push([...part]) 
            return
        } 

        for (let i = start; i < s.length; i ++) {
            if (this.isValid(s, start, i)) {
                part.push(s.substring(start, i + 1)) 
                this.dfs(i + 1, s, part, res) 
                part.pop()
            }
        }
    } 

    isValid(s, l ,r) {
        while(l < r) {
            if (s[l] !== s[r]) {
                return false
            } 

            l ++ 
            r --
        } 

        return true
    }
}
